### PR TITLE
Updated Hamiltonian Generation code so it's less terrible.

### PR DESCRIPTION
**Pull request by andygrace227**
My code now uses files to work instead of attempting to use a bunch of OpenMPI calls. It generates the Hamiltonian on up to 3 nodes, saves them to files, then solves for eigenvalues with PETSC. It does NOT save the values yet on anything other than stdout; I have yet to implement that code.

 We can scrap the agrace2_work directory as everything we need should be in the matrix_generation directory.

### CODE REVIEW

**Review by andygrace227 — COMMENTED** [review]
Unsure why these files have persisted, the gitignore should have wiped them out

**Comment by szaman19** [comment]
Do git rm or git rm --cached

File: matrix_generation/IsingHamiltonian.cpp [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+IsingHamiltonian::IsingHamiltonian(long latice_size, int which){
+    latice_size_one_dimension = latice_size;
+    N = latice_size * latice_size;
+    initializePauliMatrices();
+    std::string JMatFileName = "JMat_" + std::to_string(latice_size) + ".dynamicmatrix";
+    std::string BzMatFileName = "BzMat_" + std::to_string(latice_size) + ".dynamicmatrix";
+    std::string BxMatFileName = "BxMat_" + std::to_string(latice_size) + ".dynamicmatrix";
+    std::ifstream jfile(JMatFileName);
+    std::ifstream bzfile(BzMatFileName);
+    std::ifstream bxfile(BxMatFileName);
```
Comment (szaman19) [inline matrix_generation/IsingHamiltonian.cpp:15]: Is this still needed if the Matrices are generated within seconds?

File: matrix_generation/main.cpp [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+                ss >> indexStr;
+                ss >> valueStr;
+                long indexRaw = std::stol(indexStr);
+                value = std::stod(valueStr);
+                extractedRow = indexRaw / cols;
+                extractedColumn = indexRaw % cols;
+            }
+            MPI_Bcast(&extractedRow, 1, MPI_INT, 0, MPI_COMM_WORLD);
+            MPI_Bcast(&extractedColumn, 1, MPI_INT, 0, MPI_COMM_WORLD);
+            MPI_Bcast(&value, 1, MPI_DOUBLE, 0, MPI_COMM_WORLD);
```
Comment (szaman19) [inline matrix_generation/main.cpp]: Why broadcast? I thought the matrices were distributed across the nodes?

File: matrix_generation/makefile [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+msg_start:
+	@echo
+	@echo Starting Compilation...
+
+loadmods:
+	module load gnu8/8.3.0
+	module load openmpi3/3.1.4
+	module load petsc/3.12.0
+	module load slepc/3.12.0
```
Comment (szaman19) [inline matrix_generation/makefile]: Duplicate?
```suggestion

```

File: matrix_generation/main.cpp [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+        Argument 1: Lattice Size
+        Argument 2: J
+        Argument 3: Bx
+        Argument 4: Bz
+        Argument 5 and 6: 
+    */
+
+    if(argc < 5 || argc > 7){
+        PetscPrintf(MPI_COMM_WORLD, "This program requires 5 - 7 arguments.\n");
+        PetscPrintf(MPI_COMM_WORLD, "Usage: ./sohg <Lattice Size in 1D> <J multiplier> <Bx Multiplier> <Bz Multiplier> <--save-file or --force-gen>\n");
```
Comment (szaman19) [inline matrix_generation/main.cpp]:
```suggestion
        PetscPrintf(MPI_COMM_WORLD, "Usage: ./dhogs <Lattice Size in 1D> <J multiplier> <Bx Multiplier> <Bz Multiplier> <--save-file or --force-gen>\n");
```

**Review by szaman19 — APPROVED** [review]

File: matrix_generation/slurm.sh [hunk 5]
```diff
@@ -1,12 +1,13 @@
...
-#SBATCH --ntasks=18                   # Number of MPI tasks (i.e. processes)
+#SBATCH --ntasks=16                   # Number of MPI tasks (i.e. processes)
 #SBATCH --cpus-per-task=1            # Number of cores per MPI task 
-#SBATCH --nodes=6                    # Maximum number of nodes to be allocated
-#SBATCH --ntasks-per-node=6          # Maximum number of tasks on each node
-#SBATCH --ntasks-per-socket=6        # Maximum number of tasks on each socket
+#SBATCH --nodes=8                    # Maximum number of nodes to be allocated
+#SBATCH --acctg-freq=task=2
+#SBATCH --ntasks-per-node=4         # Maximum number of tasks on each node
+#SBATCH --ntasks-per-socket=4        # Maximum number of tasks on each socket
```
Comment (szaman19) [inline matrix_generation/slurm.sh]: @andygrace227 So I'm a little confused about this. Ideally, we want to minimize the number of processes. Inter-process communication is a killer. So ideally we want 1 or 2 (depending on the number of sockets and affinity) processes per node and do multi-threading parallelism in each node. Using message passing in a single node is not advisable unless absolutely necessary. 

Is it possible to run SLEPc in such a way, where it uses multi-threading and MPI?

File: matrix_generation/main.cpp (hunk 2 above)
Comment (andygrace227) [reply matrix_generation/main.cpp]: I had some old code in there that I forgot to take out, my bad. Tomorrow I will do another pull request with that out. My branch is properly sanitized and has that stuff out of there.

File: .vscode/settings.json [hunk 6]
```diff
@@ -1 +1,5 @@
-
+{
+    "files.associations": {
+        "fstream": "cpp"
+    }
+}
```
Comment (szaman19) [inline .vscode/settings.json]: delete
```suggestion
```

File: matrix_generation/main.cpp [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+    PetscScalar im, re;
+    EPS solver;
+    int nconv;
+
+    /* Initialize vectors */
+    MatCreateVecs(Sum, NULL, &imaginary);
+    MatCreateVecs(Sum, NULL, &real);    
+    EPSCreate(PETSC_COMM_WORLD, &solver);
+    EPSSetOperators(solver, Sum, NULL);
+    EPSSetProblemType(solver, EPS_NHEP);
```
Comment (szaman19) [inline matrix_generation/main.cpp]: The Hamiltonian is Hermitian
```suggestion
    EPSSetProblemType(solver, EPS_HEP);
```

File: matrix_generation/slurm.sh [hunk 8]
```diff
@@ -15,9 +16,9 @@
 echo "Date              = $(date)"
 echo "Hostname          = $(hostname -s)"
 echo "Working Directory = $(pwd)"
-echo ""
+echo ""T
 echo "Number of Nodes Allocated      = $SLURM_JOB_NUM_NODES"
 echo "Number of Tasks Allocated      = $SLURM_NTASKS"
 echo "Number of Cores/Task Allocated = $SLURM_CPUS_PER_TASK"
```
Comment (szaman19) [inline matrix_generation/slurm.sh:23]:
```suggestion
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$PETSC_DIR/$PETSC_ARCH/lib:$SLEPC_DIR/$PETSC_ARCH/lib
```

File: matrix_generation/slurm.sh [hunk 9]
```diff
@@ -15,9 +16,9 @@
...
 
-srun --mpi=pmix_v2 dhogs 4 1 1 1 --verbose
\ No newline at end of file
+srun --mpi=pmix_v2 dhogs 4 1 1 1
```
Comment (szaman19) [inline matrix_generation/slurm.sh]:
```suggestion
mpiexec dhogs 4 1 1 1
```

File: final_gen_solve/makefile [hunk 10]
```diff
@@ -5,9 +5,17 @@ SHELL:=/bin/bash
...
+dirs: 
+	mkdir generated_eigenvectors
+	mkdir slurm_logs
+
 matgen: pure_petsc_mat_gen.cpp genmethods.cpp vectorlib.cpp checkmethods.cpp
 	mpic++ $(FLAGS) pure_petsc_mat_gen.cpp -o $@ $(INCLUDES) $(MYLIBS)
 	@echo Finished Compilation.
 
+gencsv: generateCSV.cpp
+	gcc $(FLAGS) generateCSV.cpp -o gencsv
```
Comment (szaman19) [inline final_gen_solve/makefile]:
```
gcc $(FLAGS) generateCSV.cpp -o gencsv
```

This should be 

```
g++ $(FLAGS) generateCSV.cpp -o gencsv
```

**Review by szaman19 — CHANGES_REQUESTED** [review]

File: matrix_generation/IsingHamiltonian.cpp (hunk 1 above)
Comment (andygrace227) [reply matrix_generation/IsingHamiltonian.cpp:15]: This code is no longer used.

**Pull request merged**